### ISSUE TITLE
Support Readiness and Liveness Probe for Dapr Runtime

### ISSUE THREAD

**Issue by shalabhms** [dapr/dapr#1288]
In a scenario,  where user app is ready to serve but Dapr sidecar is having issue to come up or taking time to start then requests will start failing even when the pod is ready to serve.

This will become critical when the app is running in production, and there is need to scale the pod.  User requests might start coming to new ready pod but the requests might start failing when calling Dapr as Dapr indeed is not ready.

To avoid this scenario, Dapr need to provide a way for readiness and liveness probe so user app
can "deterministically" declare that Dapr is ready and so ready to serve requests.

> /area runtime

**Comment by shalabhms** [comment]
cc @yaron2 , @youngbupark , @amanbha ,

Let me know if this capability already exists. If not, please chime in your opinion on this feature.

**Comment by artursouza** [comment]
I was thinking about this when working on Java SDK. I was considering having a method in the SDK to wait for Dapr's sidecar health (listening on HTTP and GRPC port).

Maybe the sidecar can have a health API and the Apps (and SDKs) can probe for that. Health is not only in the startup but also help during operation.

**Comment by yaron2** [comment]
Yeah we need a formal `healthz` endpoint for our API. (HTTP only is enough here).

Outside of the scope of this issue, we need readiness and liveness probes for:

* Placement Service
* Sidecar Injector
* Operator

This is for upgrade purposes. Once those probes are in, we can ensure zero downtime upgrades.

**Comment by shalabhms** [comment]
yes, proposal is to have a readiness API and liveness API that can be probed.

**Comment by shalabhms** [comment]
> Yeah we need a formal `healthz` endpoint for our API. (HTTP only is enough here).
> 
> Outside of the scope of this issue, we need readiness and liveness probes for:
> 
> * Placement Service
> * Sidecar Injector
> * Operator
> * Sentry
> 
> This is for upgrade purposes. Once those probes are in, we can ensure zero downtime upgrades.

Agree, we need for system services as well which is very well needed to ensure zero downtime upgrades.

**Comment by youngbupark** [comment]
Closed the existing issue #725
